dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -84,174 +82,238 @@
 
-        if ( hasNestedStructure( bundle ) )
+        // Extract all enrollments and relationships, and set parent reference.
+        for ( TrackedEntity te : dataBundle.getTrackedEntities() )
         {
-            flattenPayload( bundle );
-        }
+            updateTrackedEntityReferences( te );
+            trackedEntityMap.put( te.getTrackedEntity(), te );
 
-        return bundle;
-    }
+            extractEnrollments( te )
+                .forEach( enrollment -> enrollmentHashMap.put( enrollment.getEnrollment(), enrollment ) );
 
-    private void flattenPayload( TrackerBundleParams bundle )
-    {
+            extractRelationships( te )
+                .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
+        }
 
-        List<Event> events = new ArrayList<>();
+        // Set UID for all enrollments and notes
+        dataBundle.getEnrollments().stream()
+            .peek( enrollment -> updateEnrollmentReferences( enrollment, enrollment.getTrackedEntity() ) )
+            .forEach( enrollment -> enrollmentHashMap.put( enrollment.getEnrollment(), enrollment ) );
 
-        List<Enrollment> enrollments = new ArrayList<>();
+        // Extract all events and relationships, and set parent references
+        for ( Enrollment enrollment : enrollmentHashMap.values() )
+        {
+            extractEvents( enrollment )
+                .forEach( event -> eventHashMap.put( event.getEvent(), event ) );
 
-        Set<Relationship> relationships = new HashSet<>();
+            extractRelationships( enrollment )
+                .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
 
-        // Iterate over **all** enrollments
-        for ( Enrollment enrollment : bundle.getTrackedEntities().stream()
-            .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() ) )
-        {
-            // collect all events from enrollments and add them to the flattened events collection
-            events.addAll( enrollment.getEvents().stream().map( e -> addParent( e, enrollment.getEnrollment() ) )
-                .collect( Collectors.toList() ) );
-
-            // Extract relationships from enrollment
-            relationships.addAll( enrollment.getRelationships() );
-
-            // Extract relationships from events
-            relationships
-                .addAll( events.stream().flatMap( l -> l.getRelationships().stream() ).collect( Collectors.toSet() ) );
-
-            // remove events and relationships from enrollment
-            enrollment.setEvents( Collections.emptyList() );
-            enrollment.setRelationships( Collections.emptyList() );
-            // remove relationships from events
-            events.forEach( e -> e.setRelationships( Collections.emptyList() ) );
+            enrollment.setNotes( enrollment.getNotes().stream()
```

Comment:
As we discussed, make sure that we only generate UID for `event.setEvent()`
The `trackedEntity` and `enrollment` property should be left empty if it's not present in the original payload.

```
Event:
UpdateReference(event.getEvent)
event.getEnrollment -> isEmpty() ? null : enrollment
event.getTei -> isEmpty()? null : tei
```